server/src/services/documentSymbol/onDocumentSymbol.ts

```diff
@@ -0,0 +1,122 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+/* eslint-disable @typescript-eslint/strict-boolean-expressions */
+import { DocumentSymbolParams } from "vscode-languageserver/node";
+import { DocumentSymbol, SymbolInformation } from "vscode-languageserver-types";
+import { analyze } from "@nomicfoundation/solidity-analyzer";
+import semver from "semver";
+import _ from "lodash";
+import { Language } from "@nomicfoundation/slang/language";
+import { ProductionKind } from "@nomicfoundation/slang/kinds";
+import { Cursor } from "@nomicfoundation/slang/cursor";
+import { RuleNode } from "@nomicfoundation/slang/cst";
+import { ServerState } from "../../types";
+import { SymbolTreeBuilder } from "./SymbolTreeBuilder";
+import { StructDefinition } from "./visitors/StructDefinition";
+import { StructMember } from "./visitors/StructMember";
+import { InterfaceDefinition } from "./visitors/InterfaceDefinition";
+import { FunctionDefinition } from "./visitors/FunctionDefinition";
+import { ContractDefinition } from "./visitors/ContractDefinition";
+import { EventDefinition } from "./visitors/EventDefinition";
+import { StateVariableDeclaration } from "./visitors/StateVariableDeclaration";
+import { VariableDeclaration } from "./visitors/VariableDeclaration";
+import { ConstantDefinition } from "./visitors/ConstantDefinition";
+import { ConstructorDefinition } from "./visitors/ConstructorDefinition";
+import { EnumDefinition } from "./visitors/EnumDefinition";
+import { ErrorDefinition } from "./visitors/ErrorDefinition";
+import { FallbackFunctionDefinition } from "./visitors/FallbackFunctionDefinition";
+import { LibraryDefinition } from "./visitors/LibraryDefinition";
+import { ModifierDefinition } from "./visitors/ModifierDefinition";
+import { ReceiveFunctionDefinition } from "./visitors/ReceiveFunctionDefinition";
+import { UserDefinedValueTypeDefinition } from "./visitors/UserDefinedValueTypeDefinition";
+import { SymbolVisitor } from "./SymbolVisitor";
+
+export function onDocumentSymbol(serverState: ServerState) {
+  return async (
+    params: DocumentSymbolParams
+  ): Promise<DocumentSymbol[] | SymbolInformation[] | null> => {
+    const { telemetry, solcVersions } = serverState;
+    return telemetry.trackTimingSync("onDocumentSymbol", (transaction) => {
+      const { uri } = params.textDocument;
+
+      // Find the file in the documents collection
+      const document = serverState.documents.get(uri);
+
+      if (document === undefined) {
+        throw new Error(`${uri} not found in documents`);
+      }
+
+      const text = document.getText();
+
+      // Get the document's solidity version
+      let span = transaction.startChild({ op: "solidity-analyzer" });
+      const { versionPragmas } = analyze(text);
+      span.finish();
+
+      versionPragmas.push("<= 0.8.19"); // latest supported by slang
```

Comment:
suggesgt moving this logic (Slang version selection/latest supported version) from both editor features to `slangHelpers.ts`

Reply:
Additionally, because of the two implementations, when we don't find a suitable version:

- In document symbols, we throw an error, failing the request.
- In semantic highlighting, we log an error, and return an empty response instead.

I agree that we should log this, in order to to be notified when it happens. But realistically speaking, that is not always actionable. For example, the user might be user a version that is too old (we don't plan on supporting anything before `0.4.11`) or too new (hasn't been released yet, or we are still working on supporting it). In that case, should it be still an error? should we just log the fact, and still make sure we return an empty response?

What do you think?

Reply:
Sounds great on moving to slangHelpers and logging + empty response on document symbols

Reply:
I see semantic highlighting is still throwing errors instead of catching it? Is this intended?